app/gui2/shared/retry.ts

```diff
@@ -49,15 +60,55 @@ export async function exponentialBackoff<T, E>(
       onSuccess(retries)
       return result
     } catch (error) {
-      if (retries >= maxRetries) throw error
+      if (retries >= maxRetries) {
+        onFailure(error as E, retries)
+        throw error
+      }
       if (onBeforeRetry(error as E, retries, maxRetries, delay) === false) throw error
       await wait(delay)
     }
   }
 }
 
-export const onBeforeRetry: (message: string) => BackoffOptions<any>['onBeforeRetry'] =
-  (message) => (error, retryCount, maxRetries, delay) => {
-    console.error(message + ` (${retryCount}/${maxRetries} retries), retrying after ${delay}ms...`)
+export function onBeforeRetry(
+  description: string,
+): NonNullable<BackoffOptions<any>['onBeforeRetry']> {
+  return (error, retryCount, maxRetries, delay) => {
+    console.error(
+      'Could not ' +
+        description +
+        ` (${retryCount}/${maxRetries} retries), retrying after ${delay}ms...`,
+    )
+    console.error(error)
+  }
+}
+
+export function onFailure(description: string): NonNullable<BackoffOptions<any>['onFailure']> {
+  return (error, retryCount) => {
+    console.error(
+      'Could not ' + description + ` (${retryCount}/${retryCount} retries), throwing error.`,
+    )
     console.error(error)
   }
+}
+
+export function onSuccess(description: string): NonNullable<BackoffOptions<any>['onSuccess']> {
+  return (retryCount) => {
+    if (retryCount === 0) return
+    console.info(
+      'Successfully ' +
+        description +
+        ` after ${retryCount} ${retryCount === 1 ? 'failure' : 'failures'}.`,
+    )
+  }
+}
+
+/** @param successDescription Should be in past tense, without an initial capital letter.
+ * @param successDescription Should be in present tense, without an initial capital letter. */
+export function exponentialBackoffMessages(successDescription: string, errorDescription: string) {
```

Comment:
Same @param name. And I would rename it to something like `printingCallbacks`.

Reply:
hmm... these aren't callbacks though

Reply:
> In [computer programming](https://en.wikipedia.org/wiki/Computer_programming), a callback or callback function is any reference to [executable code](https://en.wikipedia.org/wiki/Executable_code) that is passed as an [argument](https://en.wikipedia.org/wiki/Argument_(computer_science)) to another piece of code

So here we make functions, and set them as options for "retry" function. What is it if not callback? Or I misunderstand this code greatly.

Reply:
ahhhh, you mean rename the function, not the parameters

Reply:
Oh, my fault, my sentence wasn't precise. Yes, `it` meant `function`, not `parameter`.